router/acme/responder.go

```diff
@@ -0,0 +1,164 @@
+package acme
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"strings"
+	"sync"
+
+	acme "github.com/eggsampler/acme/v3"
+	ct "github.com/flynn/flynn/controller/types"
+	discoverd "github.com/flynn/flynn/discoverd/client"
+	"github.com/flynn/flynn/pkg/httphelper"
+	router "github.com/flynn/flynn/router/types"
+	"github.com/inconshreveable/log15"
+)
+
+const (
+	responderAppName     = "router"
+	responderServiceName = "router-acme-responder"
+	defaultResponderPort = "8080"
+)
+
+// Responder responds to ACME authz challenges
+type Responder struct {
+	client ControllerClient
+
+	challengeMtx sync.RWMutex
+	challenges   map[string]map[string]*acme.Challenge
+
+	ln net.Listener
+	hb discoverd.Heartbeater
+
+	log log15.Logger
+}
+
+// NewResponder returns a responder that responds to authz challenges using the
+// given controller and discoverd clients, listening for http-01 challenges at
+// the given httpAddr
+func NewResponder(client ControllerClient, discoverdClient *discoverd.Client, httpAddr string, log log15.Logger) (*Responder, error) {
+	log.Info("starting responder HTTP server", "addr", httpAddr)
+	ln, err := net.Listen("tcp", httpAddr)
+	if err != nil {
+		log.Error("error starting responder HTTP server", "err", err)
+		return nil, err
+	}
+	hb, err := discoverdClient.AddServiceAndRegisterInstance(responderServiceName, &discoverd.Instance{
+		Addr:  fmt.Sprintf(":%d", ln.Addr().(*net.TCPAddr).Port),
+		Proto: "http",
+	})
+	if err != nil {
+		log.Error("error starting responder HTTP server", "err", err)
+		return nil, err
+	}
+	r := &Responder{
+		client:     client,
+		challenges: make(map[string]map[string]*acme.Challenge),
+		ln:         ln,
+		hb:         hb,
+		log:        log,
+	}
+	handler := httphelper.ContextInjector("acme-responder", httphelper.NewRequestLogger(r))
+	go http.Serve(ln, handler)
+	return r, nil
+}
+
+// RespondHTTP01 responds to a http-01 ACME challenge by creating a route for
+// the given certificate's domain pointing at the responder's HTTP handler,
+// completing the challenge and then removing the route
+func (r *Responder) RespondHTTP01(cert *ct.ManagedCertificate, challenge *acme.Challenge, complete func() error) error {
+	route, err := r.createChallengeRoute(cert, challenge)
+	if err != nil {
+		return err
+	}
+	if err := complete(); err != nil {
+		return err
+	}
+	return r.deleteChallengeRoute(route, challenge)
```

Comment:
Should this be deferred so it gets cleaned up if `complete()` errors? We may want some kind of GC function for these.

Reply:
We could have a GC function as a recurring task.